heroku-24/setup.sh

```diff
@@ -4,37 +4,17 @@ set -euxo pipefail
 
 export DEBIAN_FRONTEND=noninteractive
 
-# This is the default `ubuntu.sources` from both the AMD64 and ARM64 images
-# combined, with `noble-backports`, `restricted` and `multiverse` removed.
-cat >/etc/apt/sources.list.d/ubuntu.sources <<EOF
-Types: deb
-URIs: http://archive.ubuntu.com/ubuntu/
-Suites: noble noble-updates
-Components: main universe
-Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg
-Architectures: amd64
-
-Types: deb
-URIs: http://security.ubuntu.com/ubuntu/
-Suites: noble-security
-Components: main universe
-Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg
-Architectures: amd64
-
-Types: deb
-URIs: http://ports.ubuntu.com/ubuntu-ports/
-Suites: noble noble-updates
-Components: main universe
-Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg
-Architectures: arm64
-
-Types: deb
-URIs: http://ports.ubuntu.com/ubuntu-ports/
-Suites: noble-security
-Components: main universe restricted multiverse
-Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg
-Architectures: arm64
-EOF
+# Disable APT repositories that contain packages with potentially problematic licenses.
+# We edit the existing file rather than using our own static file contents, since the repository
+# URIs vary across architectures, so we would otherwise have to hardcode multiple file variants.
+# There are multiple repository configuration lines in the file, all of form:
+# `Components: main universe restricted multiverse`
+# See: https://manpages.ubuntu.com/manpages/noble/en/man5/sources.list.5.html
+for repository_to_disable in multiverse restricted; do
+  # sed doesn't support lookbehind so we instead have to match against the line prefix too
+  # and then preserve it using `\1` in the replacement.
+  sed --in-place --regexp-extended "s/(Components:.*) ?${repository_to_disable}/\1/g" /etc/apt/sources.list.d/ubuntu.sources
```

Comment:
Minor - For a line like `Components: main universe restricted multiverse` this seems to leave an extra space when replacing a value in the middle of the list.  E.g.; replacing `restricted` gives the following:

```
Components: main universe  multiverse
```

It doesn't seem to affect the output in a major way since both components removed are trailing elements but it could leave extra spaces if they weren't. The modification below accounts for this:

```suggestion
  sed --in-place --regexp-extended "s/(Components:.*) ?${repository_to_disable} ?/\1/g" /etc/apt/sources.list.d/ubuntu.sources
```

Reply:
The leading ` ?` was intended to avoid that, but it seems like the `.*` being a greedy match can steal the space due to the optionality of the ` ?`. I'll make the leading space none optional instead.

Reply:
Adding the second ` ?` misses another case:

```shell
$ echo -e 'Components: keep1 remove-me keep2#\nComponents: keep1 keep2 remove-me#\nComponents:remove-me keep1 keep2#' | sed --regexp-extended "s/(Components:.*) ?remove-me ?/\1/g"
Components: keep1 keep2#
Components: keep1 keep2 #
Components:keep1 keep2#
```

I tried making ``.*`` lazy matching by changing it to `.*?`, however, because the ` ?` is optional, the `.*?` still steals the space:

```
$ echo -e 'Components: keep1 remove-me keep2#\nComponents: keep1 keep2 remove-me#\nComponents:remove-me keep1 keep2#' | sed --regexp-extended "s/(Components:.*?) ?remove-me/\1/g"
Components: keep1  keep2#
Components: keep1 keep2 #
Components: keep1 keep2#
```

Given that both the spec and the original APT sources list uses spaces after the `Components:` prefix, I think it would be safe to just use this:

```console
$ echo -e 'Components: keep1 remove-me keep2#\nComponents: keep1 keep2 remove-me#\nComponents:remove-me keep1 keep2#' | sed --regexp-extended "s/(Components:.*) remove-me/\1/g"
Components: keep1 keep2#
Components: keep1 keep2#
Components:remove-me keep1 keep2#
```

Reply:
Fixed in https://github.com/heroku/base-images/pull/287/commits/fbddf963e403b167bd01a511f690636212d66f8f